File: java/expression_evaluator/src/main/java/com/qe/test/expr/Evaluator.java
package com.qe.test.expr;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

public class Evaluator {
    private final ShuntingYard shuntingYard = new ShuntingYard();

    public double evaluate(String expression) {
        List<String> rpn = shuntingYard.toRpn(expression);
        return evaluateRpn(rpn);
    }

    public double evaluateRpn(List<String> rpn) {
        Deque<Double> stack = new ArrayDeque<>();

        for (String token : rpn) {
            switch (token) {
                case "+" -> {
                    double b = stack.pop();
                    double a = stack.pop();
                    stack.push(a + b);
                }
                case "-" -> {
                    double b = stack.pop();
                    double a = stack.pop();
                    stack.push(a - b);
                }
                case "*" -> {
                    double b = stack.pop();
                    double a = stack.pop();
                    stack.push(a * b);
                }
                case "/" -> {
                    double b = stack.pop();
                    double a = stack.pop();
                    if (b == 0.0) {
                        throw new ArithmeticException("Division by zero");
                    }
                    stack.push(a / b);
                }
                case "^" -> {
                    double b = stack.pop();
                    double a = stack.pop();
                    stack.push(Math.pow(a, b));
                }
                case "%" -> {
                    double b = stack.pop();
                    double a = stack.pop();
                    if (b == 0.0) {
                        throw new ArithmeticException("Modulo by zero");
                    }
                    stack.push(a % b);
                }
                case "_" -> stack.push(-stack.pop());
                default -> {
                    try {
                        stack.push(Double.parseDouble(token));
                    } catch (NumberFormatException e) {
                        throw new IllegalArgumentException("Unknown token: " + token, e);
                    }
                }
            }
        }

        if (stack.size() != 1) {
            throw new IllegalStateException("Invalid expression: stack has " + stack.size() + " elements");
        }

        return stack.pop();
    }
}

File: go/expression_evaluator/evaluator.go
package expression_evaluator

import (
	"fmt"
	"math"
	"strconv"
)

// Evaluate evaluates an infix arithmetic expression and returns the float64
// result. Supported operators: +, -, *, /, ^, %, parentheses, and unary minus.
func Evaluate(expr string) (float64, error) {
	postfix, err := InfixToPostfix(expr)
	if err != nil {
		return 0, fmt.Errorf("infix to postfix: %w", err)
	}
	return EvaluateRPN(postfix)
}

// EvaluateRPN evaluates a Reverse Polish Notation expression.
func EvaluateRPN(tokens []string) (float64, error) {
	var stack []float64

	for _, tok := range tokens {
		if isNumber(tok) {
			val, err := strconv.ParseFloat(tok, 64)
			if err != nil {
				return 0, fmt.Errorf("invalid number %q", tok)
			}
			stack = append(stack, val)
			continue
		}

		if tok == "_" {
			if len(stack) < 1 {
				return 0, fmt.Errorf("not enough operands for unary minus")
			}
			stack[len(stack)-1] = -stack[len(stack)-1]
			continue
		}

		if !IsOperator(tok) {
			return 0, fmt.Errorf("unknown token %q", tok)
		}

		if len(stack) < 2 {
			return 0, fmt.Errorf("not enough operands for %q", tok)
		}
		b := stack[len(stack)-1]
		a := stack[len(stack)-2]
		stack = stack[:len(stack)-2]

		var result float64
		switch tok {
		case "+":
			result = a + b
		case "-":
			result = a - b
		case "*":
			result = a * b
		case "/":
			if b == 0 {
				return 0, fmt.Errorf("division by zero")
			}
			result = a / b
		case "%":
			if b == 0 {
				return 0, fmt.Errorf("modulo by zero")
			}
			result = math.Mod(a, b)
		case "^":
			result = math.Pow(a, b)
		default:
			return 0, fmt.Errorf("unknown operator %q", tok)
		}
		stack = append(stack, result)
	}

	if len(stack) != 1 {
		return 0, fmt.Errorf("invalid expression: %d values remain on stack", len(stack))
	}
	return stack[0], nil
}

func isNumber(s string) bool {
	_, err := strconv.ParseFloat(s, 64)
	return err == nil
}
